
<img src="https://raw.githubusercontent.com/rhatiro/Curso_EBAC-Profissao_Cientista_de_Dados/main/ebac-course-utils/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Profissão: Cientista de Dados**
### **Módulo 13** | Regressão II | Exercício 1

Aluno [Roberto Hatiro Nishiyama](https://www.linkedin.com/in/rhatiro/)<br>
Data: 29 de março de 2023.

---

# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [1]:
import numpy as np
import pandas as pd

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [2]:
filepath = 'https://raw.githubusercontent.com/rhatiro/Curso_EBAC-Profissao_Cientista_de_Dados/main/M%C3%B3dulo%2013%20-%20Regress%C3%A3o%20II/database/previsao_de_renda.csv'
df = pd.read_csv(filepath_or_buffer=filepath)
df

Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  \
0               0  2015-01-01       15056    F             False   
1               1  2015-01-01        9968    M              True   
2               2  2015-01-01        4312    F              True   
3               3  2015-01-01       10639    F             False   
4               4  2015-01-01        7064    M              True   
...           ...         ...         ...  ...               ...   
14995       14995  2016-03-01       16006    F             False   
14996       14996  2016-03-01        3722    F             False   
14997       14997  2016-03-01        6194    F              True   
14998       14998  2016-03-01        4922    M              True   
14999       14999  2016-03-01       14651    F             False   

       posse_de_imovel  qtd_filhos        tipo_renda           educacao  \
0                 True           0        Empresário         Secundário   
1                 True           0       Assalariado  Superior completo   
2                 True           0        Empresário  Superior completo   
3                 True           1  Servidor público  Superior completo   
4                False           0       Assalariado         Secundário   
...                ...         ...               ...                ...   
14995             True           0        Empresário         Secundário   
14996             True           0       Pensionista  Superior completo   
14997             True           0       Assalariado  Superior completo   
14998            False           0        Empresário  Superior completo   
14999             True           2  Servidor público         Secundário   

      estado_civil tipo_residencia  idade  tempo_emprego  \
0         Solteiro            Casa     26       6.602740   
1           Casado            Casa     28       7.183562   
2           Casado            Casa     35       0.838356   
3           Casado            Casa     30       4.846575   
4         Solteiro   Governamental     33       4.293151   
...            ...             ...    ...            ...   
14995     Solteiro            Casa     48      13.887671   
14996     Solteiro            Casa     57            NaN   
14997       Casado            Casa     45       7.832877   
14998       Casado            Casa     36       4.298630   
14999        União            Casa     34       7.219178   

       qt_pessoas_residencia     renda  
0                        1.0   8060.34  
1                        2.0   1852.15  
2                        2.0   2253.89  
3                        3.0   6600.77  
4                        1.0   6475.97  
...                      ...       ...  
14995                    1.0   7990.58  
14996                    1.0  10093.45  
14997                    2.0    604.82  
14998                    2.0   3352.27  
14999                    4.0   1160.99  

[15000 rows x 15 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 
    

---

#### 1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
- Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
- Mantenha sempre a categoria mais frequente como casela de referência
- Avalie os parâmetros e veja se parecem fazer sentido prático.

In [4]:
# Analisando a frequência das categorias em cada variável:

for coluna in df.columns[3:]:
    print(coluna, '\n', df[coluna].value_counts(), '\n')

sexo 
 F    10119
M     4881
Name: sexo, dtype: int64 

posse_de_veiculo 
 False    9140
True     5860
Name: posse_de_veiculo, dtype: int64 

posse_de_imovel 
 True     10143
False     4857
Name: posse_de_imovel, dtype: int64 

qtd_filhos 
 0     10376
1      3037
2      1376
3       185
4        17
7         5
14        2
5         2
Name: qtd_filhos, dtype: int64 

tipo_renda 
 Assalariado         7633
Empresário          3508
Pensionista         2582
Servidor público    1268
Bolsista               9
Name: tipo_renda, dtype: int64 

educacao 
 Secundário             8895
Superior completo      5335
Superior incompleto     579
Primário                165
Pós graduação            26
Name: educacao, dtype: int64 

estado_civil 
 Casado      10534
Solteiro     1798
União        1078
Separado      879
Viúvo         711
Name: estado_civil, dtype: int64 

tipo_residencia 
 Casa             13532
Com os pais        675
Governamental      452
Aluguel            194
Estúdio             83
Comu

In [5]:
y, X = patsy.dmatrices(formula_like='''
                                    np.log(renda) ~ sexo 
                                                    + posse_de_veiculo 
                                                    + posse_de_imovel 
                                                    + C(qtd_filhos, Treatment(0))
                                                    + C(tipo_renda, Treatment('Assalariado'))
                                                    + C(educacao, Treatment('Secundário')) 
                                                    + C(estado_civil, Treatment('Casado')) 
                                                    + C(tipo_residencia, Treatment('Casa')) 
                                                    + idade 
                                                    + tempo_emprego 
                                                    + C(qt_pessoas_residencia, Treatment(2))
                                    ''', 
                       data=df)

In [6]:
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.358
Model:                            OLS   Adj. R-squared:                  0.357
Method:                 Least Squares   F-statistic:                     203.5
Date:                Wed, 29 Mar 2023   Prob (F-statistic):               0.00
Time:                        13:30:48   Log-Likelihood:                -13559.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12392   BIC:                         2.745e+04
Df Model:                          34                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.0932      0.039    183.905      0.000       7.018       7.169
sexo[T.M]                                                       0.7878      0.015     53.664      0.000       0.759       0.817
posse_de_veiculo[T.True]                                        0.0436      0.014      3.079      0.002       0.016       0.071
posse_de_imovel[T.True]                                         0.0828      0.014      5.907      0.000       0.055       0.110
C(qtd_filhos, Treatment(0))[T.1]                               -0.2501      0.112     -2.239      0.025      -0.469      -0.031
C(qtd_filhos, Treatment(0))[T.2]                               -0.5697      0.223     -2.559      0.011      -1.006      -0.133
C(qtd_filhos, Treatment(0))[T.3]                               -0.5560      0.486     -1.144      0.253      -1.509       0.397
C(qtd_filhos, Treatment(0))[T.4]                                0.2091      0.429      0.488      0.626      -0.631       1.049
C(qtd_filhos, Treatment(0))[T.5]                                0.1385      0.410      0.338      0.736      -0.666       0.943
C(qtd_filhos, Treatment(0))[T.7]                                0.0221      0.162      0.137      0.891      -0.295       0.339
C(qtd_filhos, Treatment(0))[T.14]                              -0.4630      0.261     -1.772      0.076      -0.975       0.049
C(tipo_renda, Treatment('Assalariado'))[T.Bolsista]             0.2259      0.241      0.937      0.349      -0.247       0.699
C(tipo_renda, Treatment('Assalariado'))[T.Empresário]           0.1547      0.015     10.359      0.000       0.125       0.184
C(tipo_renda, Treatment('Assalariado'))[T.Pensionista]         -0.2891      0.243     -1.190      0.234      -0.765       0.187
C(tipo_renda, Treatment('Assalariado'))[T.Servidor público]     0.0551      0.022      2.476      0.013       0.011       0.099
C(educacao, Treatment('Secundário'))[T.Primário]                0.0130      0.072      0.181      0.856      -0.128       0.154
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.1168      0.142      0.821      0.412      -0.162       0.396
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.1069      0.014      7.678      0.000       0.080       0.134
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0290      0.032     -0.897      0.370      -0.092       0.034
C(estado_civil, Treatment('Casado'))[T.Separado]                0.3176      0.112      2.831      0.005       0.098       0.537
C(estado_civil, Treatment('Casado'))[T.Solteiro]                0.2556      0.110      2.319      0.020       0.040       0.472
C(estado_ci

#### 2. Remova a variável menos significante e analise:
- Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
- Observe os parâmetros e veja se algum se alterou muito.

In [7]:
# Removendo a variável `tipo_residencia` com o menor P>|t| == 0.200:

smf.ols(formula='''
                np.log(renda) ~ sexo 
                                + posse_de_veiculo 
                                + posse_de_imovel 
                                + C(qtd_filhos, Treatment(0))
                                + C(tipo_renda, Treatment('Assalariado'))
                                + C(educacao, Treatment('Secundário')) 
                                + C(estado_civil, Treatment('Casado')) 
                                + idade 
                                + tempo_emprego 
                                + C(qt_pessoas_residencia, Treatment(2))
                ''', 
        data=df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.358
Model:                            OLS   Adj. R-squared:                  0.357
Method:                 Least Squares   F-statistic:                     238.5
Date:                Wed, 29 Mar 2023   Prob (F-statistic):               0.00
Time:                        13:30:50   Log-Likelihood:                -13561.
No. Observations:               12427   AIC:                         2.718e+04
Df Residuals:                   12397   BIC:                         2.741e+04
Df Model:                          29                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.0973      0.038    187.689      0.000       7.023       7.171
sexo[T.M]                                                       0.7896      0.015     53.937      0.000       0.761       0.818
posse_de_veiculo[T.True]                                        0.0436      0.014      3.084      0.002       0.016       0.071
posse_de_imovel[T.True]                                         0.0817      0.014      5.943      0.000       0.055       0.109
C(qtd_filhos, Treatment(0))[T.1]                               -0.2492      0.112     -2.231      0.026      -0.468      -0.030
C(qtd_filhos, Treatment(0))[T.2]                               -0.5694      0.223     -2.558      0.011      -1.006      -0.133
C(qtd_filhos, Treatment(0))[T.3]                               -0.5605      0.486     -1.153      0.249      -1.513       0.392
C(qtd_filhos, Treatment(0))[T.4]                                0.1888      0.428      0.441      0.659      -0.650       1.028
C(qtd_filhos, Treatment(0))[T.5]                                0.1192      0.410      0.291      0.771      -0.684       0.922
C(qtd_filhos, Treatment(0))[T.7]                                0.0214      0.162      0.133      0.894      -0.295       0.338
C(qtd_filhos, Treatment(0))[T.14]                              -0.4639      0.261     -1.776      0.076      -0.976       0.048
C(tipo_renda, Treatment('Assalariado'))[T.Bolsista]             0.2271      0.241      0.942      0.346      -0.246       0.700
C(tipo_renda, Treatment('Assalariado'))[T.Empresário]           0.1551      0.015     10.401      0.000       0.126       0.184
C(tipo_renda, Treatment('Assalariado'))[T.Pensionista]         -0.2917      0.243     -1.202      0.230      -0.768       0.184
C(tipo_renda, Treatment('Assalariado'))[T.Servidor público]     0.0563      0.022      2.534      0.011       0.013       0.100
C(educacao, Treatment('Secundário'))[T.Primário]                0.0087      0.072      0.121      0.904      -0.132       0.149
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.1153      0.142      0.810      0.418      -0.164       0.394
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.1066      0.014      7.674      0.000       0.079       0.134
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0287      0.032     -0.891      0.373      -0.092       0.034
C(estado_civil, Treatment('Casado'))[T.Separado]                0.3182      0.112      2.837      0.005       0.098       0.538
C(estado_civil, Treatment('Casado'))[T.Solteiro]                0.2567      0.110      2.330      0.020       0.041       0.473
C(estado_ci

**Resposta:**

Após remover a variável `tipo_residencia`, o modelo não apresentou melhora nem piora significativa. O coeficiente de determinação manteve o mesmo valor de 0.358 e o *p-value* das variáveis teve uma variação muito pequena, praticamente insignificante.

#### 3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.

In [8]:
smf.ols(formula='''
                np.log(renda) ~ sexo 
                                + posse_de_veiculo 
                                + posse_de_imovel
                                + tipo_renda == 'Empresário'
                                + tipo_renda == 'Servidor público'
                                + educacao == 'Superior completo'
                                + qt_pessoas_residencia == 3
                                + qt_pessoas_residencia == 4
                                + idade 
                                + tempo_emprego
                ''', 
        data=df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     688.2
Date:                Wed, 29 Mar 2023   Prob (F-statistic):               0.00
Time:                        13:30:51   Log-Likelihood:                -13576.
No. Observations:               12427   AIC:                         2.717e+04
Df Residuals:                   12416   BIC:                         2.726e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                    7.0815      0.035    201.800      0.000       7.013       7.150
sexo[T.M]                                    0.7844      0.015     53.961      0.000       0.756       0.813
posse_de_veiculo[T.True]                     0.0437      0.014      3.117      0.002       0.016       0.071
posse_de_imovel[T.True]                      0.0824      0.014      6.011      0.000       0.055       0.109
tipo_renda == 'Empresário'[T.True]           0.1534      0.015     10.318      0.000       0.124       0.182
tipo_renda == 'Servidor público'[T.True]     0.0572      0.022      2.578      0.010       0.014       0.101
educacao == 'Superior completo'[T.True]      0.1083      0.014      7.979      0.000       0.082       0.135
qt_pessoas_residencia == 3[T.True]           0.0443      0.017      2.596      0.009       0.011       0.078
qt_pessoas_residencia == 4[T.True]           0.0843      0.022      3.806      0.000       0.041       0.128
idade                                        0.0060      0.001      7.783      0.000       0.004       0.008
tempo_emprego                                0.0615      0.001     59.297      0.000       0.059       0.064
==============================================================================
Omnibus:                        0.888   Durbin-Watson:                   2.022
Prob(Omnibus):                  0.641   Jarque-Bera (JB):                0.862
Skew:                           0.018   Prob(JB):                        0.650
Kurtosis:                       3.018   Cond. No.                         241.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Resposta:**

Ao comparar o modelo final com o modelo inicial, os indicadores não apresentaram nenhuma melhora ou piora significativa. O coeficiente de determinação diminuiu de 0.358 para 0.357 e o AIC de 27190 para 27170.

---